test/smartwallet/rifSmartWallet.test.ts

```diff
@@ -0,0 +1,920 @@
+import { BaseProvider } from '@ethersproject/providers';
+import {
+  CustomSmartWallet__factory,
+  SmartWallet__factory,
+} from '@rsksmart/rif-relay-contracts';
+import chai, { expect } from 'chai';
+import chaiAsPromised from 'chai-as-promised';
+import { ethers as hardhat } from 'hardhat';
+import { SignerWithAddress } from '@nomiclabs/hardhat-ethers/signers';
+import { TestForwarder, TestTarget } from 'typechain-types';
+import { Wallet, providers, Contract } from 'ethers';
+import {
+  TEST_TOKEN_NAME,
+  NON_REVERT_TEST_TOKEN_NAME,
+  TETHER_TOKEN_NAME,
+  INITIAL_SMART_WALLET_TOKEN_AMOUNT,
+  RifSmartWallet,
+  TokenToTest,
+  prepareToken,
+  mintTokens,
+  getLogArguments,
+} from './rifSmartWalletUtils';
+import {
+  createSmartWalletFactory,
+  createRifSmartWallet,
+  createEnvelopingRequest,
+  getSuffixDataAndSignature,
+  getSuffixData,
+} from '../utils/TestUtils';
+import {
+  RelayRequest,
+  INTERNAL_TRANSACTION_ESTIMATED_CORRECTION,
+} from '@rsksmart/rif-relay-client';
+import {
+  getLocalEip712Signature,
+  TypedRequestData,
+} from '../utils/EIP712Utils';
+import SmartWalletJson from '../../artifacts/@rsksmart/rif-relay-contracts/contracts/smartwallet/SmartWallet.sol/SmartWallet.json';
+import CustomSmartWalletJson from '../../artifacts/@rsksmart/rif-relay-contracts/contracts/smartwallet/CustomSmartWallet.sol/CustomSmartWallet.json';
+import { CustomSmartWallet, SmartWallet } from '@rsksmart/rif-relay-contracts';
+import nodeConfig from 'config';
+
+chai.use(chaiAsPromised);
+
+const INITIAL_SMART_WALLET_RBTC_AMOUNT = 50;
+const TOKEN_AMOUNT_TO_TRANSFER = 1;
+const RBTC_AMOUNT_TO_TRANSFER = hardhat.utils.parseEther('1');
+
+const CUSTOM_SMART_WALLET_TYPE: TypeOfWallet = 'CustomSmartWallet';
+const SMART_WALLET_TYPE: TypeOfWallet = 'SmartWallet';
+const TYPES_OF_WALLETS: TypeOfWallet[] = [
+  CUSTOM_SMART_WALLET_TYPE,
+  SMART_WALLET_TYPE,
+];
+
+const TOKENS = [TEST_TOKEN_NAME, NON_REVERT_TEST_TOKEN_NAME, TETHER_TOKEN_NAME];
+
+const IS_DEPLOY_REQUEST = false;
+
+const CONFIG_BLOCKCHAIN = 'blockchain';
+const CONFIG_RSK_URL = 'rskNodeUrl';
+
+const RSK_URL = nodeConfig.get<string>(
+  `${CONFIG_BLOCKCHAIN}.${CONFIG_RSK_URL}`
+);
+
+const isCustomSmartWallet = (smartWalletType: string) =>
+  smartWalletType === CUSTOM_SMART_WALLET_TYPE;
+
+type TypeOfWallet = 'CustomSmartWallet' | 'SmartWallet';
+
+function createRifSmartWalletFromABI(
```

Comment:
why do we need to use a sw from ABI?

Reply:
Hardhat is not providing the error attached to some reverts.